### ISSUE TITLE
the necessity of spliting the raw data into chunks?

### ISSUE THREAD

**Issue by a550461053** [mravanelli/pytorch-kaldi#158]
I find that the cost time of reading one chunk data is more than training time. Can we explain the necessity of spliting the raw data into many chunks? Why not use generator and iterator to load data？

**Comment by TParcollet** [comment]
Hi! In the current version, the data loading pipeline is not optimal. It is multi-threaded, but not with a PyTorch generator. Pytorch-Kaldi will soon be transformed into a much bigger and powerful tool, that will integrates faster data pipelines. To reduce this problem, you can reduce the number of chunks (it will increase the VRAM consumption).

**Comment by mravanelli** [comment]
Hi,
let me add some other comments/suggestions:

1- If you are using a cluster it is really very important to copy the
dataset locally and read from the local machine. Usually shared filesystems
can be very slow or unstable.
2- We suggest to have chunk of 1 hours of speech maximum.
3- One issue might arise when we have to process with the current code a
huge dataset. The issue is that, due to kaldi, we read all the labels of
all the dataset for every chunk. If is much better to read all the labels
once and keep them in memory.
4- The reason why we are using our dataloader and not the pytorch one is
that in the early version of pytorch (e.g, 0.3/0.4) we have noticed some
weird behaviours with the pytorch one and we decided to implement our ones
customized on kaldi features/labels. Now the dataload should be more stable
and we think to include it in the next release.
5- When reading data there are two options. One is to read a lot of data in
rather large chunks, store the chunks in memory and the process everything
chunk by chunk. An alternative  is the directly read from disk the signal
for every batch. For speech application, the first solution, however, is
more suitable especially for feedforwad neural networks. In the latter
case, in fact, we want batches that are very "diverse" and composed of many
different speakers (when I read a sentence I have to read all the features
of that sentence, and I cannot read just one time frame in isolation for
instance. This limits the possible diversity).
6- As said by Titouan, we are working for a radically new version of the
toolkit. There will be an announcement in the next days.

Best,

Mirco

**Comment by jprobichaud** [comment]
I'm not sure where to ask for this, but here seems like a good place: we are trying to use pytorch-kaldi with lots of audio (our small dataset for training is 5k hours) and since it's an issue with the current setup, we're looking forward for this update.  My questions are the following: 

1. will the setup of experiments (.cfg files) remain somewhat relevant in this new version or will the specification of an experiment be completely different
1. what about the .proto files?  There isn't a tDNN implementation directly available and I'm looking to see if I should try to make one
1. by "next days", how close are we from this release?  Weeks or months?  

Thanks for these great codebases!

**Comment by mravanelli** [comment]
Hi,
thank you veru much for the interest into the project.

it shouldn't be a problem training on a very large amount of data. The only
thing is that the dataloader might be not efficient due to the high number
of labels. My suggestion is to read all the labels at the beginning of the
script and keep them in memory (this way one can rad only the feature part
in the dataloader). The new version will do something like this.

1- We are totally revising and extending the toolkit. There will be an
announcement soon. The config file will change substantially to make it
feasible dealing with other speech tasks such as speaker verification,
speech enhancement, etc.
2- TDNN and other models will be available. If you are able to make a
working TDNN model, it could be great if you can share it!
3- More than a new version, it will be a new project. It will take months,
but we will continue meanwhile maintaining the current version of
pytorch-kaldi

**Comment by jprobichaud** [comment]
I'm a little confused: will the soon-to-be announcement correspond to the public release of the new project (accessible to the public) or will it take many months before that new codebase gets released?  

Basically, I'm trying to see if it's worth investing in the current pytorch-kaldi project or if we should wait few days/weeks and jump on the new project?

**Comment by mravanelli** [comment]
The project that will extend pytorch-kaldi will be only announced (we are
looking for collaborators to achieve the goals with have in mind). The new
project will be a very ambitious initiative that will take months before
having a working code. In the meanwhile, we will maintain and provide
support for the current pytorch-kaldi project as well.

Mirco


<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by a550461053** [comment]
Thanks for every one! I have been busy with training so much that I don't have time to respond to you. I was training the ASR-kaldi system by one machine with 8 gpus and will training the ASR-kaldi system by clusters. In response to this problem, I use a trick way to change the method of training and data process that pre-process two chunks data in two threads while training one chunk data. Under that conditions the time of pre-processing one chunk is t1 and training one chunk cost time t2( t2 < t1, 2 * t2 > t1). Then, the all training-cost time is t2 * num_chunk * num_epoch.
